velox/functions/sparksql/DateTimeFunctions.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/DateTimeFunctions.h"
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+std::optional<Timestamp> makeTimeStampFromDecodedArgs(
+    vector_size_t row,
+    DecodedVector* yearVector,
+    DecodedVector* monthVector,
+    DecodedVector* dayVector,
+    DecodedVector* hourVector,
+    DecodedVector* minuteVector,
+    DecodedVector* microsVector) {
+  // Check hour.
+  auto hour = hourVector->valueAt<int32_t>(row);
+  if (hour < 0 || hour > 24) {
+    return std::nullopt;
+  }
+  // Check miniute.
+  auto minute = minuteVector->valueAt<int32_t>(row);
+  if (minute < 0 || minute > 60) {
+    return std::nullopt;
+  }
+  // Check microseconds.
+  auto micros = microsVector->valueAt<int64_t>(row);
+  if (micros < 0) {
+    return std::nullopt;
+  }
+  auto seconds = micros / util::kMicrosPerSec;
+  if (seconds > 60 || (seconds == 60 && micros % util::kMicrosPerSec != 0)) {
+    return std::nullopt;
+  }
+
+  // year, month, day will be checked in utils::daysSinceEpochFromDate;
+  try {
+    auto daysSinceEpoch = util::daysSinceEpochFromDate(
+        yearVector->valueAt<int32_t>(row),
+        monthVector->valueAt<int32_t>(row),
+        dayVector->valueAt<int32_t>(row));
+    // micros has at most 8 digits (2 for seconds + 6 for microseconds),
+    // thus it's safe to cast micros from int64_t to int32_t.
+    auto localMicros = util::fromTime(hour, minute, 0, (int32_t)micros);
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxUserError& e) {
+    return std::nullopt;
```

Comment:
To use try-catch for each row may have poor performace. Maybe we can take use of Status to represent the computing outcome of `daysSinceEpochFromDate` method.

Reply:
Maybe we can optimize that in a separate PR. @mbasmanova Do you have any suggestion?

Reply:
@rui-mo Rui has a good point. We should not throw and catch exceptions per row. See https://velox-lib.io/blog/optimize-try_cast.